packages/modules/kafka/src/kafka-container.ts

```diff
@@ -83,35 +112,83 @@ export class KafkaContainer extends GenericContainer {
+        KAFKA_PROCESS_ROLES: "broker,controller",
+        KAFKA_CONTROLLER_QUORUM_VOTERS: `${this.environment["KAFKA_BROKER_ID"]}@${network}:${KAFKA_CONTROLLER_PORT}`,
+        KAFKA_CONTROLLER_LISTENER_NAMES: "CONTROLLER",
+      });
     }
 
-    command += "echo '' > /etc/confluent/docker/ensure \n";
-    command += "/etc/confluent/docker/run \n";
-    this.withCommand(["sh", "-c", command]);
+    // Change the wait strategy to wait for a log message from a fake starter script
+    // so that we can put a real starter script in place at that moment
+    this.originalWaitinStrategy = this.waitStrategy;
+    this.waitStrategy = Wait.forLogMessage(WAIT_FOR_SCRIPT_MESSAGE);
+    this.withEntrypoint(["sh"]);
+    this.withCommand([
+      "-c",
+      `echo '${WAIT_FOR_SCRIPT_MESSAGE}'; while [ ! -f ${STARTER_SCRIPT} ]; do sleep 0.1; done; ${STARTER_SCRIPT}`,
+    ]);
   }
 
   public override async start(): Promise<StartedKafkaContainer> {
+    if (this.mode === KafkaMode.KRAFT && this.saslSslConfig && this.isLessThanCP(7, 5)) {
+      throw new Error(
+        `Provided Confluent Platform's version ${this.imageName.tag} is not supported in Kraft mode with sasl (must be ${MIN_KRAFT_SASL_VERSION} or above)`
+      );
+    }
     return new StartedKafkaContainer(await super.start());
   }
 
   protected override async containerStarted(
     container: StartedTestContainer,
     inspectResult: InspectResult
   ): Promise<void> {
-    await this.updateAdvertisedListeners(container, inspectResult);
-    if (this.saslSslConfig) {
+    let command = "#!/bin/bash\n";
+    const advertisedListeners = this.updateAdvertisedListeners(container, inspectResult);
+    // exporting KAFKA_ADVERTISED_LISTENERS with the container hostname
+    command += `export KAFKA_ADVERTISED_LISTENERS=${advertisedListeners}\n`;
+
+    if (this.mode !== KafkaMode.KRAFT || this.isLessThanCP(7, 4)) {
+      // Optimization: skip the checks
+      command += "echo '' > /etc/confluent/docker/ensure \n";
+    }
+    if (this.mode === KafkaMode.KRAFT) {
+      if (this.saslSslConfig) {
+        command += this.commandKraftCreateUser(this.saslSslConfig);
+      }
+      if (this.isLessThanCP(7, 4)) {
+        command += this.commandKraft();
+      }
+    } else if (this.mode === KafkaMode.EMBEDDED_ZOOKEEPER) {
+      command += this.commandZookeeper();
+    }
+
+    // Run the original command
+    command += "/etc/confluent/docker/run \n";
+    container.copyContentToContainer([{ content: command, target: STARTER_SCRIPT, mode: 0o777 }]);
+
+    const client = await getContainerRuntimeClient();
+    const dockerContainer = client.container.getById(container.getId());
+    const boundPorts = BoundPorts.fromInspectResult(client.info.containerRuntime.hostIps, inspectResult).filter(
+      this.exposedPorts
+    );
+    await waitForContainer(client, dockerContainer, this.originalWaitinStrategy, boundPorts);
```

Comment:
This is different, to manually invoke a wait strategy after a container is said to have started. Ideally we should get this working just by setting `this.waitStrategy` and nothing else. E.g would it work if on line 139 we do 
```ts
this.waitStrategy = Wait.forAll([
  this.waitStrategy, 
  Wait.forLogMessage(WAIT_FOR_SCRIPT_MESSAGE)
]);
```

Reply:
This won't work unfortunately.

The main problem is we need to gather some info which is only available in when container is started and copy it to the container. Unlike the java implementation, some of this methods are not available on not-started container. It can be achieved, but the code will be messier and will have some duplicated functionality. I'll try to prepare a commit with that implementation.

Reply:
OK, we can look at this some other time. For now I think it's more important to add support for Kraft